### PR TITLE
feat: e2e tests

### PR DESCRIPTION
**Pull request by enicio**

### CODE REVIEW

File: .mocharc.js [hunk 1]
```diff
@@ -7,5 +7,5 @@ module.exports = {
     reporter: 'mochawesome',
     'reporter-option': ['reportDir=test-results'],
     require: 'source-map-support/register',
-    spec: ['test/out/**/*.test.js']
+    // spec: ['test/out/**/*.test.js']
```
Comment (lmalheiro) [inline .mocharc.js:10]: Isn't better to remove it?

File: package.json [hunk 2]
```diff
@@ -12,8 +12,9 @@
     "cover": "nyc --reporter=lcov --reporter=text npm t",
     "exec": "node dist/main.js",
     "lint": "npx eslint --fix src && npx eslint --fix test",
-    "test": "rimraf ./test/out && npm run build && npm run build:tests && mocha",
-    "rebuild:all": "npm run clean && npm run build:all"
+    "test": "rimraf ./out && npm run build && npm run build:tests && mocha ./out/ts/*.test.js",
+    "rebuild:all": "npm run clean && npm run build:all",
+    "e2e": "rimraf ./out && npm run build && npm run build:tests && mocha ./out/e2e-test/*.test.js"
```
Comment (lmalheiro) [inline package.json:17]: I believe we might be missing a full cleanup of the S3 storage. If yes, we could leave it to be done in in a another PR to allow us to close this one now.

File: test/e2e-test/nextjs.test.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        const dataToString = data.toString();
+        const dataToArray = dataToString.split('\n');
+        dataToArray.splice(1,1); //Remove import of NextJS/Image
+        dataToArray.splice(61,1,'               <img src="/vercel.svg" alt="Vercel Logo" width="72" height="16" />\n');
+        const arrayToText =  dataToArray.join('\n');
+        fs.writeFileSync(`${template}/pages/index.js`,arrayToText)
+        process.chdir(template);
+        await execFile(`npx next build`);
+        await execFile(`npx next export`);
+    });
```
Comment (lmalheiro) [inline test/e2e-test/nextjs.test.ts:62]: Shouldn't we see at least one `expect` here? Perhaps a test on the `execFile` return code?

File: test/e2e-test/nextjs.test.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            response.Contents?.forEach(obj => filesOnS3Bucket.push(obj.Key?.replace('__static_content_test/nextjs-e2e/', '')));
+
+            if (response.NextContinuationToken) {
+                bucketParams.ContinuationToken = response.NextContinuationToken;
+                await getAllKeys(bucketParams, filesOnS3Bucket);
+            }
+            return filesOnS3Bucket;
+        }
+
+        await getAllKeys(bucketParams, filesOnS3Bucket);
```
Comment (lmalheiro) [inline test/e2e-test/nextjs.test.ts:113]: It seems this test is relying on the previous test output. Probably both tests should be made independent or merged. To speedup  the process, I favor the merge.

File: test/e2e-test/flareact.test.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            return filesOnS3Bucket;
+        }
+
+        await getAllKeys(bucketParams, filesOnS3Bucket);
+
+        const manifest = fs.readFileSync(`${template}/worker/manifest.json`);
+        const manifestArray = Object.values(JSON.parse(manifest.toString()));
+        const filesOnS3BucketContainManifestFiles = manifestArray.every((each) =>  filesOnS3Bucket.includes(each) );
+        expect(filesOnS3BucketContainManifestFiles).to.be.true;
+    });
```
Comment (lmalheiro) [inline test/e2e-test/flareact.test.ts:103]: It seems this test is relying on the previous test output. Probably both tests should be made independent or merged. To speedup the process, I favor the merge.

Also, as it seems that we don't clean the S3 before the tests, how do we know that we are not retrieving content from a previous successful test?

File: test/e2e-test/flareact-error.test.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            await execFile('npm install', {});
+            const packageLock = fs.existsSync(path.join(template,'package-lock.json'));
+            expect(packageLock).to.be.true;
+        });
+
+        it('Build the Flareact project without "azion.json" file and expect it fail', async () => {
+            const expectOnError =  "Couldn't read file 'azion.json' at the project's root directory. Because ENOENT: no such file or directory, open 'azion.json'\n"
+
+            function run(cmd: string) {
+                return new Promise((resolve, _reject) => {
```
Comment (lmalheiro) [inline test/e2e-test/flareact-error.test.ts:69]: As you did "promesify" the `child_process.exec`, you probably don't need to use it with a "callback". It should return a promise.

File: test/e2e-test/flareact-error.test.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+        before(async () => {
+            await copy(path.join(localOutput, 'test', 'config-files', 'azion-flareact-without-S3.json'),path.join(template,'azion.json'))
+        });
+
+        it('Build the Flareact project with "azion.json" without S3 credentials', async () => {
+            const expectOnError =  "S3 credentials not set either in the configuration file or as environment variables.\n"
+
+            function run(cmd: string) {
+                return new Promise((resolve, _reject) => {
```
Comment (lmalheiro) [inline test/e2e-test/flareact-error.test.ts:92]: Same observation as above.

**Review by lmalheiro — APPROVED** [review]
LGTM! 🚀

**Review by augustollenz — APPROVED** [review]

**Review by lmalheiro — APPROVED** [review]

**Review by MagnunAVFAzion — APPROVED** [review]
